### PR TITLE
Raise a plan error on union if column count is not the same between plans

### LINKED ISSUE #13092: Datafusion 42 does not raise plan error on some queries
### Describe the bug

Since datafusion 41, i got Datafusion Error when some SQL queries contain errors. 

```sql
SELECT "Product" FROM "my_table" UNION ALL SELECT "Product", "Id" FROM "my_table"
```

raises

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Cell In[4], line 4
      1 query = """
      2 SELECT "Product" FROM "my_table" UNION ALL SELECT "Product", "Id" FROM "my_table"
      3 """
----> 4 df = ctx.sql(query)

File ~/.../site-packages/datafusion/context.py:514, in SessionContext.sql(self, query, options)
    499 """Create a :py:class:`~datafusion.DataFrame` from SQL query text.
    500 
    501 Note: This API implements DDL statements such as ``CREATE TABLE`` and
   (...)
    511     DataFrame representation of the SQL query.
    512 """
    513 if options is None:
--> 514     return DataFrame(self.ctx.sql(query))
    515 return DataFrame(self.ctx.sql_with_options(query, options.options_internal))

Exception: DataFusion error: Plan("Union queries must have the same number of columns, (left is 1, right is 2)")
```

With new datafusion version, this doesn't happens anymore (it crashes later, when i run `.collect()` after the .sql() result)

### To Reproduce

```py
import pyarrow as pa
import datafusion as df
import pyarrow.dataset as pda

t = pa.Table.from_pydict(
    {
        "Id": [0, 0, 0, 0, 0, 1, 1, 1, 1],
        "Product": ["A", "A", "A", "B", "C", "B", "C", "B", "B"],
    },
)

ctx = df.SessionContext()
ctx.register_dataset(name="my_table", dataset=pda.dataset(t))

query = """
SELECT "Product" FROM "my_table" UNION ALL SELECT "Product", "Id" FROM "my_table"
"""
df = ctx.sql(query)  # with DF 41 this riases an error, with 42 doesn't
```

### Expected behavior

DataFusion error exception correctly raised

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #13092

## Rationale for this change

Raise the error earlier, restore the behaviour of previous versions of DF

## What changes are included in this PR?

code, slt test.

## Are these changes tested?

yes

## Are there any user-facing changes?
no

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -1482,6 +1482,15 @@ pub fn validate_unique_names<'a>(
 /// [`TypeCoercionRewriter::coerce_union`]: https://docs.rs/datafusion-optimizer/latest/datafusion_optimizer/analyzer/type_coercion/struct.TypeCoercionRewriter.html#method.coerce_union
 /// [`coerce_union_schema`]: https://docs.rs/datafusion-optimizer/latest/datafusion_optimizer/analyzer/type_coercion/fn.coerce_union_schema.html
 pub fn union(left_plan: LogicalPlan, right_plan: LogicalPlan) -> Result<LogicalPlan> {
+    if left_plan.schema().fields().len() != right_plan.schema().fields().len() {
+        return plan_err!(
+            "UNION queries have different number of columns: \
+            left has {} columns whereas right has {} columns",
```
Comment (findepi) [inline datafusion/expr/src/logical_plan/builder.rs:1488]: This is good check to have. 👍

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
This is good check to have. 👍 

Obviously having same column count doesn't guarantee compatibility. The types need to be coercible, so normally the column count check should go together with the type check.
Reading rationale for https://github.com/apache/datafusion/pull/11961 it seems that we have two+ phases:

1. "loose" Logical Plans where types may not match
   - this is apparently the phase what plan builder operates at, because it directly supports dataframe frontend, I believe
   - then we do "type coercion optimizer" (which is not really an optimizer)
2. "strict" Logical Plans where types are expected to match

The consequence of this is that 

- it's unclear what guarantees are provided at which phase, because phases do not stand out in the code
- we can still "type coercion optimizer" and other coercion related functions (like expr.get_type) on strict Logical Plans which cannot benefit from them

I would want to see these addressed, that's the idea of https://github.com/apache/datafusion/issues/12723. 

Maybe we should have a form of validator of logical plan well-formedness (union branches match, the types match _exactly_) and the check would go there.

The downside of such approach (and of current state as well) is that, when you have code taking Logical Plan UNION, you can't really assume much (things may or may not have been checked), so the code needs to be very defensively written. 
Wonder whether we can leverage Rust's compiler type system to help us with this.

cc @alamb @ozankabak @jonahgao

**Comment by alamb** [comment]
> Maybe we should have a form of validator of logical plan well-formedness (union branches match, the types match exactly) and the check would go there.

I think this is what the the type coercion analyzer pass already does. 

If the goal is to provide users errors sooner (when DataFrame is created) rather than later (when DataFrame is run) some other ideas:
1. Add a function to DataFrame (`DataFrame::optimize`, `DataFrame::validate`, `DataFrame::check` that runs)
2. Add some configuration / option that runs coercion logic automatically on each DataFrame call

**Comment by alamb** [comment]
> Reading rationale for https://github.com/apache/datafusion/pull/11961 it seems that we have two+ phases:

I agree with this analysis and conclusion

> type coercion optimizer" (which is not really an optimizer)

I agree it is not an optimizer -- that is why it an `Analyzer`  [`datafusion::optimizer::analyzer::type_coercion::TypeCoercion` ](https://docs.rs/datafusion/latest/datafusion/optimizer/analyzer/type_coercion/struct.TypeCoercion.html)

I think the analyzer docs explain this: https://docs.rs/datafusion/latest/datafusion/optimizer/trait.AnalyzerRule.html

File: datafusion/expr/src/logical_plan/builder.rs (hunk 1 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:1488]: 👍  I agree -- it is an easy check to add that can catch clearly incorrect plans (though as @findepi  notes it won't catch all issues)

**Review by alamb — APPROVED** [review]
Thanks @Omega359 and @findepi 

cc @jonahgao

**Review by jonahgao — APPROVED** [review]
Thanks @Omega359 . Raising error earlier makes sense to me

**Pull request merged**